File: rxhttp/src/main/java/rxhttp/wrapper/entity/ExpandOutputStream.java
package rxhttp.wrapper.entity;

import android.content.Context;
import android.net.Uri;

import org.jetbrains.annotations.NotNull;

import java.io.File;
import java.io.FileNotFoundException;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.OutputStream;

/**
 * User: ljx
 * Date: 2022/9/27
 * Time: 17:16
 */
public final class ExpandOutputStream<T> extends OutputStream {

    private final T expand;
    private final OutputStream os;

    public ExpandOutputStream(T expand, OutputStream os) {
        this.expand = expand;
        this.os = os;
    }

    public static ExpandOutputStream<String> open(File file, boolean append) throws FileNotFoundException {
        return new ExpandOutputStream<>(file.getAbsolutePath(), new FileOutputStream(file, append));
    }

    public static ExpandOutputStream<Uri> open(Context context, Uri uri, boolean append) throws FileNotFoundException {
        OutputStream os = context.getContentResolver().openOutputStream(uri, append ? "wa" : "w");
        return new ExpandOutputStream<>(uri, os);
    }

    public T getExpand() {
        return expand;
    }

    @Override
    public void write(int b) throws IOException {
        os.write(b);
    }

    @Override
    public void write(@NotNull byte[] b) throws IOException {
        os.write(b);
    }

    @Override
    public void write(@NotNull byte[] b, int off, int len) throws IOException {
        os.write(b, off, len);
    }

    @Override
    public void close() throws IOException {
        os.close();
    }

    @Override
    public void flush() throws IOException {
        os.flush();
    }
}

File: rxhttp/src/main/java/rxhttp/wrapper/entity/ExpandOutputStream.kt
package rxhttp.wrapper.entity

import android.content.Context
import android.net.Uri
import java.io.File
import java.io.FileOutputStream
import java.io.OutputStream

/**
 * @param expand When the download is complete, the object is returned
 * @param os Download OutputStream
 */
class ExpandOutputStream<out T>(
    val expand: T,
    val os: OutputStream
) {
    override fun toString(): String = "($expand, $os)"
}

internal fun File.toOutputStream(append: Boolean = false) =
    ExpandOutputStream(absolutePath, FileOutputStream(this, append))

internal fun Uri.toOutputStream(
    context: Context,
    append: Boolean = false
): ExpandOutputStream<Uri> {
    val os = context.contentResolver.openOutputStream(this, if (append) "wa" else "w")
    return ExpandOutputStream(this, os)
}